sdk/bare/common/sys/Logging-Binary-Format.md

```diff
@@ -0,0 +1,40 @@
+# Logging Binary Format
+
+## Overview
+
+The AMDC can dump the logged variables to the UART serial console using either text-based formatting or the binary
+format. This document describes the binary formatting which appears on the UART interface.
+
+## Dumping Variables
+
+Each logged variable must be dumped one at a time (i.e. if the user has 10 slots, each must be dumped one by one).
+
+To start a dump of the Nth variable slot: `log dump bin N`
+
+## Binary Format
+
+All data is little-endian.
+
+### Magic Header (16 bytes)
+
+0x12345678 (repeated four times)
+
+### Number of Samples (4 bytes, integer)
+
+0x00000010
+
+### Data Type (4 bytes, integer)
+
+- 0x00000001 = int
+- 0x00000002 = float
+- 0x00000003 = double
+
+### Each Sample (8 bytes)
+
+| Byte 0-3            | Byte 4-7 |
+|---------------------|----------|
+| Timestamp (usec)    | Data     |
```

Comment:
The timestamp field adds a lot of data overhead (2x). I wonder if we could do something creative to minimize this (to increase data throughput over UART -- and also, later when we consider real time data streaming via gigabit ethernet).

Ideas:
1. Do we need such a long time stamp? Could we reduce it down to 16 bit? 
    1. **32 bit unsigned integer** can hold a decimal of up to `4,294,967,295`. If we are logging at a rate of `1us`, this corresponds to `4,294 s` worth of data with time stamps.
    2. **16 bit unsigned integer** can hold a decimal of up to `65,535`. If we are logging at a rate of `1us`, this corresponds to `65 ms` worth of data with time stamps. I guess it seems likely that a logging run of 1 - 2 seconds may be frequently requested, but I guess that it would be okay if the time stamp field wrapped around.

2. What about having a `start time stamp` field (4 bytes) and a `sampling rate field` that is provided before the data. Since data is sampled at a fixed interval, the receiver can deal with interpreting the sample time. 

3. I guess that maybe we are concerned with recording jitter between samples. In which case, we could implement idea 2) _and_ utilize an 8 bit signed time stamp per sample. This 8 bit time stamp would indicate the time deviation from the last sample (in `us`). Positive values indicate the sample was recorded that many `us` after it was expected (based on the `sampling rate field`), negative values indicate the sample was recorded that many `us` before it was expected (again, based on the `sampling rate field`)

Reply:
This is a good idea (removing timestamp from each sample) -- will improve throughput by 2x. Nice.

The timestamp field used a "fake" timestamp (i.e. approximated) and not the actual CPU time (i.e. your 3rd point about sampling jitter does not exist in the current code). It would be nice to put in the sampling time deviation as an offset to the normalized sampling rate. However, I am going to make this a later feature. In general, FFT theory only supports fixed sampling intervals anyway, so it might not actually be helpful to have PERFECTLY precise timestamps.... I believe the sampling jitter is very small anyway (< 100ns).

I will work on your 2nd point. After chatting with @Nick-Hemenway, we think that each transmission should also include a metadata field of the sampling rate. We can assume the logging starts at _t = 0_ for all logged variables, so we don't need to transmit timestamps over UART.

Reply:
Added embedded code to support this here: 3ca8d278c708adf7bec324ddbf6dcf4071406a5e

Reply:
Embedded code doesn't output timestamps for each sample: bc3ab5118ba60707d523394feec24af04b7491c0

Reply:
Updated binary doc for no timestamp output: 2ba36445c938b9c08858b05be88b446e65ddade9

Reply:
Update python parsing to accept new binary format: bd19eaa1d8695a1dd21e27d4ec1b2d017ee1c7ae